File: scripts/diagnostics/andromeda_turkey_sync.php
<?php
error_reporting(0);ob_start();$pdo=null;$lock=null;$phase='setup';
try{
 if(PHP_SAPI!=='cli')throw new RuntimeException();
 $root=realpath(getcwd());if(!$root||basename($root)!=='anytoour.ru')throw new RuntimeException();
 $preview=$root.'/_preview/search3-anex-candidate';
 require_once $preview.'/api-andromeda-search3-preview.php';
 require_once $root.(is_file($root.'/data/db-v1.php')?'/data/db-v1.php':'/v2/data/db-v1.php');
 $pdo=v2_data_db();$pdo->setAttribute(PDO::ATTR_ERRMODE,PDO::ERRMODE_EXCEPTION);
 $config=require $preview.'/.andromeda-private.php';$private=dirname($config['catalog_path']);
 $stage=$private.'/turkey-catalog-v1';
 $request=json_decode(file_get_contents('php://stdin'),true,32,JSON_THROW_ON_ERROR);
 $local=$pdo->query("SELECT id,name FROM catalog_countries WHERE name='Турция' AND is_active=1")->fetchAll(PDO::FETCH_ASSOC);
 if(count($local)!==1)throw new RuntimeException();$country=(int)$local[0]['id'];
 $lock=fopen($private.'/turkey-catalog.lock','c');if(!$lock||!flock($lock,LOCK_EX))throw new RuntimeException();
 if($request['operation']==='capture'){
  $phase='capture_reservation';if(file_exists($stage)||!mkdir($stage,0700))throw new RuntimeException();
  anytour_andromeda_search3_save($stage.'/reservation.json',['state'=>'inflight','country_id'=>$country]);
  $saved=json_decode(file_get_contents($config['catalog_path']),true,32,JSON_THROW_ON_ERROR);
  $stateId=anytour_anex_search3_dictionary_id($saved['state']['payload']['STATE'],['Турция']);
  $departure=(int)$saved['all']['params']['TOWNFROMINC'];
  $client=new AnyTourAndromedaClient(static function($url,$options)use($private){
   anytour_andromeda_search3_budget($private);return (new AnyTourAndromedaTransport(true))($url,$options);
  },true);
  $phase='supplier_catalog';$client->login($config['username'],$config['password']);
  $params=['TOWNFROMINC'=>$departure,'STATEINC'=>$stateId];$payload=$client->catalog('all',$params);
  $bytes=json_encode($payload,JSON_UNESCAPED_UNICODE|JSON_THROW_ON_ERROR);
  foreach([$config['username'],$config['password'],rawurlencode($config['username']),rawurlencode($config['password'])] as $secret)if($secret!==''&&strpos($bytes,$secret)!==false)throw new RuntimeException();
  $saved['all']=['action'=>'all','params'=>$params,'payload'=>$payload];
  $saved['local_country_id']=$country;$saved['local_country_name']=$local[0]['name'];
  anytour_andromeda_search3_save($stage.'/catalog.json',$saved);
  $phase='local_catalog';$pdo->exec('START TRANSACTION READ ONLY');
  $h=$pdo->prepare('SELECT id,name,country_id,country_name,region_name,subregion_name,category FROM catalog_hotels WHERE country_id=? AND is_active=1 ORDER BY id LIMIT 20001');
  $h->execute([$country]);$hotels=$h->fetchAll(PDO::FETCH_ASSOC);
  $q=$pdo->prepare('SELECT a.hotel_id,a.alias FROM hotel_aliases a JOIN catalog_hotels h ON h.id=a.hotel_id WHERE h.country_id=? AND h.is_active=1 ORDER BY a.hotel_id,a.alias LIMIT 50001');
  $q->execute([$country]);$aliases=$q->fetchAll(PDO::FETCH_ASSOC);$pdo->exec('ROLLBACK');
  if(count($hotels)>20000||count($aliases)>50000)throw new RuntimeException();
  $result=['status'=>'captured','country_id'=>$country,'supplier_country_id'=>$stateId,'catalog_sha256'=>hash_file('sha256',$stage.'/catalog.json'),
   'catalog'=>$saved['all'],'local'=>['complete'=>true,'country_id'=>$country,'hotels'=>$hotels,'aliases'=>$aliases],'supplier_calls'=>2];
  anytour_andromeda_search3_save($stage.'/capture-result.json',['status'=>'captured','catalog_sha256'=>$result['catalog_sha256'],'hotels'=>count($payload['HOTELS'])]);
 }elseif($request['operation']==='import'){
  $phase='import_reservation';if(!is_file($stage.'/capture-result.json')||is_file($stage.'/import-reservation.json'))throw new RuntimeException();
  $saved=json_decode(file_get_contents($stage.'/catalog.json'),true,32,JSON_THROW_ON_ERROR);
  if(hash_file('sha256',$stage.'/catalog.json')!==$request['catalog_sha256']||$saved['local_country_id']!==$country)throw new RuntimeException();
  $catalog=[];foreach($saved['all']['payload']['HOTELS'] as $h)$catalog[(string)$h['id']]=$h;
  $rows=$request['rows'];if(count($rows)!==count($catalog))throw new RuntimeException();
  $seen=[];foreach($rows as $r){
   $id=$r['external_hotel_id'];$e=json_decode($r['evidence_json'],true,32,JSON_THROW_ON_ERROR);
   if(!is_string($id)||!isset($catalog[$id])||isset($seen[$id])||$e['source']!=$catalog[$id]||!in_array($r['decision_status'],['accepted','pending','conflict'],true))throw new RuntimeException();
   if($r['decision_status']==='accepted' ? (!is_int($r['local_hotel_id'])||$r['local_hotel_id']<1||(string)$catalog[$id]['stateKey']!==(string)$saved['all']['params']['STATEINC']) : $r['local_hotel_id']!==null)throw new RuntimeException();
   $seen[$id]=true;
  }
  anytour_andromeda_search3_save($stage.'/import-reservation.json',['state'=>'inflight','request_sha256'=>hash('sha256',json_encode($request))]);
  $phase='import';$pdo->beginTransaction();
  $before=$pdo->query('SELECT * FROM andromeda_hotel_identities ORDER BY supplier_namespace,external_hotel_id FOR UPDATE')->fetchAll(PDO::FETCH_ASSOC);
  $existing=[];foreach($before as $r)$existing[$r['supplier_namespace'].':'.$r['external_hotel_id']]=$r;
  $lookup=$pdo->prepare('SELECT name FROM catalog_hotels WHERE id=? AND country_id=? AND is_active=1');
  $insert=$pdo->prepare("INSERT INTO andromeda_hotel_identities(supplier_namespace,external_hotel_id,local_hotel_id,decision_status,catalog_sha256,evidence_sha256,evidence_json) VALUES('andromeda_catalog',?,?,?,?,?,?)");
  $counts=['accepted'=>0,'pending'=>0,'conflict'=>0];$skipped=[];$inserted=0;
  foreach($rows as $r){
   if(isset($existing['andromeda_catalog:'.$r['external_hotel_id']])){$skipped[]=$r['external_hotel_id'];continue;}
   if($r['local_hotel_id']!==null){$lookup->execute([$r['local_hotel_id'],$country]);$name=$lookup->fetchColumn();$e=json_decode($r['evidence_json'],true);if(!$name||$name!==$e['target_name'])throw new RuntimeException();}
   $insert->execute([$r['external_hotel_id'],$r['local_hotel_id'],$r['decision_status'],$request['catalog_sha256'],hash('sha256',$r['evidence_json']),$r['evidence_json']]);++$inserted;++$counts[$r['decision_status']];
  }
  $after=$pdo->query('SELECT * FROM andromeda_hotel_identities ORDER BY supplier_namespace,external_hotel_id')->fetchAll(PDO::FETCH_ASSOC);
  $afterIndex=[];foreach($after as $r)$afterIndex[$r['supplier_namespace'].':'.$r['external_hotel_id']]=$r;
  foreach($existing as $key=>$old)if($afterIndex[$key]!==$old)throw new RuntimeException();
  foreach($rows as $r){$key='andromeda_catalog:'.$r['external_hotel_id'];if(isset($existing[$key]))continue;$a=$afterIndex[$key];if((string)$a['local_hotel_id']!==(string)$r['local_hotel_id']||$a['decision_status']!==$r['decision_status']||$a['evidence_sha256']!==hash('sha256',$r['evidence_json']))throw new RuntimeException();}
  $pdo->commit();$phase='catalog_activation';
  $dir=$private.'/countries';if(!is_dir($dir)&&!mkdir($dir,0700))throw new RuntimeException();
  if(file_exists($dir.'/'.$country.'.json'))throw new RuntimeException();
  anytour_andromeda_search3_save($dir.'/'.$country.'.json',$saved);
  $result=['status'=>'imported','country_id'=>$country,'supplier_country_id'=>$saved['all']['params']['STATEINC'],
   'inserted'=>$inserted,'counts'=>$counts,'existing_preserved'=>count($before),'existing_id_collisions'=>$skipped,
   'catalog_hotels'=>count($catalog),'catalog_sha256'=>hash_file('sha256',$dir.'/'.$country.'.json'),'readback_verified'=>true,'supplier_calls'=>0];
  anytour_andromeda_search3_save($stage.'/import-result.json',$result);
 }else throw new RuntimeException();
}catch(Throwable $e){
 if($pdo instanceof PDO && $pdo->inTransaction())$pdo->rollBack();
 $result=['status'=>'failed','phase'=>$phase,'retry'=>false];
}
if($lock){flock($lock,LOCK_UN);fclose($lock);}while(ob_get_level())ob_end_clean();
echo json_encode($result,JSON_UNESCAPED_UNICODE|JSON_UNESCAPED_SLASHES);

File: scripts/diagnostics/andromeda_turkey_sync.py
"""One-shot Turkey catalog capture and append-only country identity import."""
import argparse, hashlib, json
from pathlib import Path
from andromeda_catalog_sync import names
from andromeda_hotel_candidates import save

def match(data):
    local=data['local'];catalog=data['catalog'];country=int(data['country_id']);supplier=int(data['supplier_country_id'])
    if local.get('complete') is not True or int(local['country_id'])!=country or int(catalog['params']['STATEINC'])!=supplier:
        raise ValueError('country mismatch')
    hotels={int(h['id']):h for h in local['hotels']}
    if any(int(h['country_id'])!=country for h in hotels.values()):raise ValueError('foreign local hotel')
    index={}
    for identifier,h in hotels.items():
        for key in names(h['name']):index.setdefault(key,set()).add(identifier)
    for alias in local['aliases']:
        identifier=int(alias['hotel_id'])
        if identifier not in hotels:raise ValueError('foreign alias')
        for key in names(alias['alias']):index.setdefault(key,set()).add(identifier)
    rows=[];seen=set()
    for source in catalog['payload']['HOTELS']:
        external=str(source['id'])
        if not external.isdigit() or external in seen:raise ValueError('invalid source key')
        seen.add(external);candidates=set()
        for value in [source['name'],source.get('lName','')]:
            for key in names(value):candidates.update(index.get(key,set()))
        target=None;status='pending';reason='no_unique_country_name'
        if str(source['stateKey'])!=str(supplier):status='conflict';reason='supplier_country_conflict'
        elif len(candidates)==1:target=next(iter(candidates));status='accepted';reason='unique_country_name_or_former_name'
        evidence={'source':source,'candidate_ids':sorted(candidates),'target_name':hotels[target]['name'] if target else None,'reason':reason}
        rows.append({'external_hotel_id':external,'local_hotel_id':target,'decision_status':status,
            'evidence_json':json.dumps(evidence,ensure_ascii=False,sort_keys=True,separators=(',',':'))})
    if not rows:raise ValueError('empty source catalog')
    return {'operation':'import','catalog_sha256':data['catalog_sha256'],'rows':rows}

def main():
    parser=argparse.ArgumentParser();parser.add_argument('phase',choices=['prepare','apply'])
    parser.add_argument('--directory',type=Path,required=True);args=parser.parse_args()
    root=args.directory;root.mkdir(exist_ok=True)
    import anex_search3_owner_decisions as owner
    php=Path(__file__).with_suffix('.php').read_text().removeprefix('<?php')
    if args.phase=='prepare':
        save(root/'capture-reservation.json',{'state':'inflight'},exclusive=True)
        data=owner.ssh_php(php,{'operation':'capture'},maximum_bytes=8000000)
        save(root/'capture.json',data,exclusive=True)
        if data.get('status')!='captured':raise ValueError('capture stopped; inspect without replay')
        request=match(data);save(root/'import-request.json',request,exclusive=True)
        print(json.dumps({'status':'prepared','country_id':data['country_id'],'supplier_country_id':data['supplier_country_id'],
            'source_hotels':len(request['rows']),'local_hotels':len(data['local']['hotels']),
            'counts':{s:sum(r['decision_status']==s for r in request['rows']) for s in ['accepted','pending','conflict']}}))
    else:
        request=json.loads((root/'import-request.json').read_bytes())
        save(root/'import-reservation.json',{'state':'inflight','request_sha256':hashlib.sha256((root/'import-request.json').read_bytes()).hexdigest()},exclusive=True)
        result=owner.ssh_php(php,request,maximum_bytes=8000000)
        save(root/'result.json',result,exclusive=True);print(json.dumps(result))
        if result.get('status')!='imported' or result.get('readback_verified') is not True:raise ValueError('import not confirmed; no replay')
if __name__=='__main__':main()
